samples/cpp/videocapture_with_audio.cpp

```diff
@@ -0,0 +1,148 @@
+/*
+ * This sample demonstrates how to read and display frames from a video file, a camera device or an image sequence.
+ * The sample also demonstrates extracting audio buffers from a video file.
+
+ * Capturing video and audio simultaneously from camera device is not supported by OpenCV.
+*/
+
+
+#include <opencv2/core.hpp>
+#include <opencv2/videoio.hpp>
+#include <opencv2/highgui.hpp>
+#include <iostream>
+
+using namespace cv;
+using namespace std;
+
+string keys =
+"{input i | | Path to input image or video file. Skip this argument to capture frames from a camera.}"
+"{with_audio wt | | set --with_audio if you want to extract audio buffers. If no input video file is provided only audio samples will be extracted}"
+"{help h | | print usage}";
+
+int main(int argc, char** argv)
+{
+    CommandLineParser parser(argc, argv, keys);
+
+    parser.about("This sample demonstrates how to read and display frames from a video file, a camera device or an image sequence \n"
+                 "The sample also demonstrates extracting audio buffers from a video file\n"
+                 "Capturing video and audio simultaneously from camera device is not supported by OpenCV\n"
+                 "Usage:\n --input/-i=<video file, image sequence> Skip this argument to capture frames from a camera\n"
+                 "--with_audio/-wt, set --with_audio if you want to extract audio buffers. If no input video file is provided only audio samples will be extracted\n"
+                 "q,Q,esc -- quit\n"
+                 "You can also pass the path to an image sequence and OpenCV will treat the sequence just like a video\n"
+                 "example: --input=right%02d.jpg\n");
+
+    // Check for 'help' argument to display usage information
+    if (parser.has("help"))
+    {
+        parser.printMessage();
+        return 0;
+    }
+
+    // Get input file path and check if audio extraction is required
+    string file = parser.get<string>("input");
+    bool isWithAudio = parser.has("with_audio");
+
+    Mat videoFrame;
+    Mat audioFrame;
+    vector<vector<Mat>> audioData;
+    VideoCapture cap;
+
+    // Setup VideoCapture based on user input
+    if (isWithAudio) {
+        if (file.empty()) {
+            // No input file, extract audio from default device
+            cout << "Extracting audio samples from default device..." << endl;
+            vector<int> micParams{ CAP_PROP_AUDIO_STREAM, 0, CAP_PROP_VIDEO_STREAM, -1, CAP_PROP_AUDIO_DATA_DEPTH, CV_16S };
+            cap.open(0, CAP_ANY, micParams);
+            if (!cap.isOpened()) {
+                cerr << "ERROR! Unable to open camera and microphone" << endl;
+                return -1;
+            }
+        } else {
+            // Open input file for both video and audio
+            vector<int> fileParams{ CAP_PROP_AUDIO_STREAM, 0, CAP_PROP_VIDEO_STREAM, 0, CAP_PROP_AUDIO_DATA_DEPTH, CV_16S };
+            cap.open(file, CAP_ANY, fileParams);
```

Comment:
Original example used `CAP_MSMF`, this one doesn't. For example on Linux with GStreamer it seems to hang or works veery slow.

I ran it like this:
```
./bin/example_cpp_videocapture_with_audio -i=../opencv_extra/testdata/highgui/audio/test_audio.wav --with_audio
```
Maybe it should be somehow adapted to another backend? On what platform did you test it?

`audio_spectrogram` example seem to be working with the same file.

Reply:
@mshabunin It works with a video file but fails when initialized using a camera device in GStreamer backend when run like this

`./bin/example_cpp_videocapture_with_audio --wt`

 It works on windows with MSMF backend

Reply:
More infromation:
```
vector<int> micParams{ CAP_PROP_AUDIO_STREAM, 0, CAP_PROP_VIDEO_STREAM, 0, CAP_PROP_AUDIO_DATA_DEPTH, CV_16S };
cap.open(0, CAP_ANY, micParams);
```
The above configuration does not work on linux with gstreamer. 

However, if I try gstreamer directly using the following pipeline it works
```
gst-launch-1.0 autovideosrc ! videoconvert ! autovideosink autoaudiosrc ! audioconvert ! autoaudiosink
```

Reply:
> @mshabunin The comparison to the `audio_spectogram` sample is not one-to-one because in this sample, we are trying to extract both video frames and audio, whereas in the `audio_spectogram` sample, we are only extracting audio. This sample is not a replacement to `audio_spectogram` sample.
> 
> So test needs to be done with a video file that also has an audio. I test on Linux and Windows like this and it works:
> 
> ```
> ./bin/example_cpp_videocapture_with_audio -i=../opencv_extra/testdata/highgui/video/VID00003-20100701-2204.avi --with_audio
> ```
> 
> The case where this sample does not work with CAP_ANY, CAP_MSMF or CAP_GSTREAMER is this:
> 
> ```
> ./bin/example_cpp_videocapture_with_audio  --with_audio
> ```
> 
> i.e. when we try to extract video and audio from a camera device.

@mshabunin The sample is supposed to work with video which might have an audio. Please try `/opencv_extra/testdata/highgui/videoVID00003-20100701-2204.avi `